packages/chain-mon/src/wd-mon/service.ts

```diff
@@ -0,0 +1,232 @@
+        },
+        nodeConnectionFailures: {
+          type: Gauge,
+          desc: 'Number of times node connection has failed',
+          labels: ['layer', 'section'],
+        },
+      },
+    })
+  }
+
+  async init(): Promise<void> {
+    // Connect to L1.
+    await waitForProvider(this.options.l1RpcProvider, {
+      logger: this.logger,
+      name: 'L1',
+    })
+
+    // Connect to L2.
+    await waitForProvider(this.options.l2RpcProvider, {
+      logger: this.logger,
+      name: 'L2',
+    })
+
+    this.state.messenger = new CrossChainMessenger({
+      l1SignerOrProvider: this.options.l1RpcProvider,
+      l2SignerOrProvider: this.options.l2RpcProvider,
+      l1ChainId: await getChainId(this.options.l1RpcProvider),
+      l2ChainId: await getChainId(this.options.l2RpcProvider),
+    })
+
+    // Not detected by default.
+    this.state.forgeryDetected = false
+
+    // For now we'll just start take it from the env or the tip of the chain
+    if (this.options.startBlockNumber === -1) {
+      this.state.highestUncheckedBlockNumber =
+        await this.options.l1RpcProvider.getBlockNumber()
+    } else {
+      this.state.highestUncheckedBlockNumber = this.options.startBlockNumber
+    }
+
+    this.logger.info(`starting L1 block height`, {
+      startBlockNumber: this.state.highestUncheckedBlockNumber,
+    })
+  }
+
+  // K8s healthcheck
+  async routes(router: ExpressRouter): Promise<void> {
+    router.get('/healthz', async (req, res) => {
+      return res.status(200).json({
+        ok: !this.state.forgeryDetected,
+      })
+    })
+  }
+
+  async main(): Promise<void> {
+    // Get current block number
+    let latestL1BlockNumber: number
+    try {
+      latestL1BlockNumber = await this.options.l1RpcProvider.getBlockNumber()
+    } catch (err) {
+      this.logger.error(`got error when connecting to node`, {
+        error: err,
+        node: 'l1',
+        section: 'getBlockNumber',
+      })
+      this.metrics.nodeConnectionFailures.inc({
+        chainId: this.state.messenger.l1ChainId,
+        section: 'getBlockNumber',
+      })
+      await sleep(this.options.sleepTimeMs)
+      return
+    }
+
+    // See if we have a new unchecked block
+    if (latestL1BlockNumber <= this.state.highestUncheckedBlockNumber) {
+      // The RPC provider is behind us, wait a bit
+      await sleep(this.options.sleepTimeMs)
+      return
+    }
+
+    this.logger.info(`checking recent blocks`, {
+      fromBlockNumber: this.state.highestUncheckedBlockNumber,
+      toBlockNumber: latestL1BlockNumber,
+    })
+
+    // Perform the check
+    let proofEvents: Event[]
+    try {
+      // The query includes events in the blockNumbers given as the last two arguments
+      proofEvents =
+        await this.state.messenger.contracts.l1.OptimismPortal.queryFilter(
+          this.state.messenger.contracts.l1.OptimismPortal.filters.WithdrawalProven(),
+          this.state.highestUncheckedBlockNumber,
```

Comment:
This can error if the range is too large

Reply:
Hmmm... Good call out. Perhaps in the event of an error we try a progressively smaller interval?

What do you suggest?

Reply:
I think that if we are running this against our own nodes, this is fine for now. To make this portable, i suggest that we write a utility function that is able to take a `ethers.Contract.queryFilter` function, a filter, a start, an end and an interval and then it basically makes calls the size of the interval and concats them together so it looks like a single call from the pov of the consumer of this util

Reply:
To be honest this util should be upstreamed into ethers, if there was an "options" param to `queryFilter`. But for now we would be able to use this util across all TS services and have a config param that states how large the interval is and its infinite if not passed in

Reply:
Agree it's surprising there isn't a standard way to handle this. 

> I think that if we are running this against our own nodes, this is fine for now.

SGTM. 

I believe other comments have all been addressed, is there anything outstanding required for approval?